sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/provider/GenericTranslatorProvider.java

```diff
@@ -0,0 +1,246 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.euphoria.core.translate.provider;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.CompositeOperator;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.FlatMap;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.Join;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.ReduceByKey;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.Union;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.base.Operator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.CompositeOperatorTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.FlatMapTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.JoinTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.OperatorTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.ReduceByKeyTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.TranslatorProvider;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.UnionTranslator;
+
+/**
+ * Adjustable {@link TranslatorProvider} that selects first suitable translation for the registered
+ * operator.
+ *
+ * <p>{@link OperatorTranslator Translators} can be added by calling variants of {@link
+ * GenericTranslatorProvider.Builder#register(Class, OperatorTranslator) register} method during
+ * build. Order of registration is important. Building is started by {@link #newBuilder()}.
+ */
+public class GenericTranslatorProvider implements TranslatorProvider {
+
+  private final List<TranslationDescriptor> possibleTranslators;
+
+  private GenericTranslatorProvider(List<TranslationDescriptor> possibleTranslators) {
+    this.possibleTranslators = possibleTranslators;
+  }
+
+  public static GenericTranslatorProvider createWithDefaultTranslators() {
+    return GenericTranslatorProvider.newBuilder()
+        .register(FlatMap.class, new FlatMapTranslator<>())
+        .register(Union.class, new UnionTranslator<>())
+        .register(ReduceByKey.class, new ReduceByKeyTranslator<>())
+        .register(Join.class, new JoinTranslator<>())
+        // register fallback operator translator to decompose composit operators
+        .register(op -> op instanceof CompositeOperator, new CompositeOperatorTranslator<>())
```

Comment:
I still have a feeling that CompositeOperator deserves its own provider, and the GenericTranslatorProvider would allow to register translators for specific operators only.

But I don't have any strong argument for this :)

Reply:
And I do not have any arguments against it. :-) My hope is that current MR will open doors for a such changes in future when some arguments arrive.